File: webapp/src/egp/approx.ts
/**
 * The analytic spectral approximation of Section 3, ported from egp/approx.py:
 * the classical high-resolution formula with the quantizer noise power Δ²/12
 * added to the spectrum, so spectral zeros stay finite.
 *
 * It is computed on a *midpoint* frequency grid — exact spectral nulls fall at
 * rational multiples of π, where log of a grid value would be −∞ even though
 * the integral converges — and needs no spectral floor, which makes it
 * independent of the factorization machinery: a genuinely external check on
 * the particle filter rather than a restatement of it.
 */
import { fft, nextPow2 } from './fft'

export const QUANTIZER_FLOOR_BITS = 0.5 * Math.log2((2 * Math.PI * Math.E) / 12)

/** |H(ω)|² on the midpoint grid ωₘ = π(m + ½)/n over [0, π]. */
export function psdMidpoint(kernel: Float64Array, n = 16384): Float64Array {
  const size = nextPow2(Math.max(2 * n, kernel.length))
  if (size !== 2 * n) throw new Error('psdMidpoint: n must be a power of two >= kernel length / 2')
  const re = new Float64Array(size)
  const im = new Float64Array(size)
  for (let j = 0; j < kernel.length; j++) {
    const phi = (-Math.PI / size) * j
    re[j] = kernel[j] * Math.cos(phi)
    im[j] = kernel[j] * Math.sin(phi)
  }
  fft(re, im)
  const spectrum = new Float64Array(n)
  for (let m = 0; m < n; m++) spectrum[m] = re[m] * re[m] + im[m] * im[m]
  return spectrum
}

/**
 * (1/4π) ∫ log₂( 2πe (S(ω) + Δ²/12) / Δ² ) dω, in bits/sample — the
 * closed-form reference of Section 3. Valid for Δ ≲ σ; in the coarse regime
 * it tends to the 0.2546-bit quantizer floor while the true rate tends to 0.
 */
export function entropyRateApprox(kernel: Float64Array, delta: number, n = 16384): number {
  const spectrum = psdMidpoint(kernel, n)
  const noise = (delta * delta) / 12
  let mean = 0
  for (let m = 0; m < n; m++) mean += Math.log2(spectrum[m] + noise)
  mean /= n
  return 0.5 * (Math.log2(2 * Math.PI * Math.E) + mean) - Math.log2(delta)
}

/** The part of `entropyRateApprox` above the fixed QUANTIZER_FLOOR_BITS. */
export function snrIntegral(kernel: Float64Array, delta: number, n = 16384): number {
  const spectrum = psdMidpoint(kernel, n)
  const noise = (delta * delta) / 12
  let mean = 0
  for (let m = 0; m < n; m++) mean += Math.log2(1 + spectrum[m] / noise)
  return (0.5 * mean) / n
}

File: egp/src/egp/approx.py
from __future__ import annotations

import math

import numpy as np

QUANTIZER_FLOOR = 0.5 * math.log2(2 * math.pi * math.e / 12.0)


def psd(kernel, n: int = 4096, midpoint: bool = False):
    """|H(omega)|^2 on a grid of [0, pi].

    The midpoint grid avoids spectral nulls at rational multiples of pi, where
    log S is integrable but not evaluable.
    """
    h = np.asarray(kernel, dtype=float).ravel()
    size = 2 * n
    if midpoint:
        rotated = h * np.exp(-1j * (np.pi / size) * np.arange(h.size))
        response = np.fft.fft(rotated, size)[:n]
        omega = np.pi * (np.arange(n) + 0.5) / n
    else:
        response = np.fft.rfft(h, size)
        omega = np.pi * np.arange(response.size) / n
    return omega, np.abs(response) ** 2


def innovation_variance(kernel, n: int = 16384) -> float:
    """exp((1/2pi) int log S dw), zero when S vanishes on a set of positive measure."""
    _, spectrum = psd(kernel, n, midpoint=True)
    with np.errstate(divide="ignore"):
        return float(np.exp(np.mean(np.log(spectrum))))


def entropy_rate_approx(kernel, delta: float, n: int = 16384) -> float:
    """(1/4pi) int log2( 2 pi e (S(w) + Delta^2/12) / Delta^2 ) dw, in bits/sample.

    The high-resolution formula with the quantizer noise floor Delta^2/12 added
    to the spectrum, so that spectral zeros stay finite.
    """
    h = np.trim_zeros(np.asarray(kernel, dtype=float).ravel())
    delta = float(delta)
    _, spectrum = psd(h, n, midpoint=True)
    return 0.5 * (
        math.log2(2 * math.pi * math.e)
        + float(np.mean(np.log2(spectrum + delta**2 / 12.0)))
    ) - math.log2(delta)


def snr_integral(kernel, delta: float, n: int = 16384) -> float:
    """The part of `entropy_rate_approx` above the fixed QUANTIZER_FLOOR."""
    h = np.trim_zeros(np.asarray(kernel, dtype=float).ravel())
    _, spectrum = psd(h, n, midpoint=True)
    return 0.5 * float(np.mean(np.log2(1.0 + spectrum / (float(delta) ** 2 / 12.0))))


def entropy_rate_lower_bound(kernel, delta: float) -> float:
    """(1/2) log2(1 + 2 pi e sigma_inf^2 / Delta^2)."""
    variance = innovation_variance(kernel)
    return 0.5 * math.log2(1.0 + 2 * math.pi * math.e * variance / float(delta) ** 2)
